keyboards/gmmk/numpad/numpad.c

```diff
@@ -0,0 +1,136 @@
+    {0, CS4_SW2,  CS5_SW2,  CS6_SW2  },  //  1  /
+    {0, CS7_SW1,  CS8_SW1,  CS9_SW1  },  //  2  *
+    {0, CS7_SW2,  CS8_SW2,  CS9_SW2  },  //  3  -
+    {0, CS4_SW3,  CS5_SW3,  CS6_SW3  },  //  4  7
+    {0, CS4_SW4,  CS5_SW4,  CS6_SW4  },  //  5  8
+    {0, CS7_SW3,  CS8_SW3,  CS9_SW3  },  //  6  9
+    {0, CS7_SW4,  CS8_SW4,  CS9_SW4  },  //  7  +
+    {0, CS4_SW5,  CS5_SW5,  CS6_SW5  },  //  8  4
+    {0, CS4_SW6,  CS5_SW6,  CS6_SW6  },  //  9  5
+    {0, CS7_SW5,  CS8_SW5,  CS9_SW5  },  //  10 6
+    {0, CS4_SW7,  CS5_SW7,  CS6_SW7  },  //  11 1
+    {0, CS4_SW8,  CS5_SW8,  CS6_SW8  },  //  12 2
+    {0, CS7_SW7,  CS8_SW7,  CS9_SW7  },  //  13 3
+    {0, CS7_SW8,  CS8_SW8,  CS9_SW8  },  //  14 ENTER
+    {0, CS4_SW9,  CS5_SW9,  CS6_SW9  },  //  15 0
+    {0, CS7_SW9,  CS8_SW9,  CS9_SW9  },  //  16 .
+    {0, CS1_SW1,  CS2_SW1,  CS3_SW1  },  //  17 LED18
+    {0, CS1_SW2,  CS2_SW2,  CS3_SW2  },  //  18 LED19
+    {0, CS1_SW3,  CS2_SW3,  CS3_SW3  },  //  19 LED20
+    {0, CS1_SW4,  CS2_SW4,  CS3_SW4  },  //  20 LED21
+    {0, CS1_SW5,  CS2_SW5,  CS3_SW5  },  //  21 LED22
+    {0, CS1_SW6,  CS2_SW6,  CS3_SW6  },  //  22 LED23
+    {0, CS1_SW7,  CS2_SW7,  CS3_SW7  },  //  23 LED24
+    {0, CS10_SW1, CS11_SW1, CS12_SW1 },  //  24 LED27
+    {0, CS10_SW2, CS11_SW2, CS12_SW2 },  //  25 LED28
+    {0, CS10_SW3, CS11_SW3, CS12_SW3 },  //  26 LED29
+    {0, CS10_SW4, CS11_SW4, CS12_SW4 },  //  27 LED30
+    {0, CS10_SW5, CS11_SW5, CS12_SW5 },  //  28 LED31
+    {0, CS10_SW6, CS11_SW6, CS12_SW6 },  //  29 LED32
+    {0, CS10_SW7, CS11_SW7, CS12_SW7 },  //  30 LED33
+};
+
+led_config_t g_led_config = {{
+    { 0,   1,   2,   3  },
+    { 4,   5,   6,   7  },
+    { 8,   9,   10,  NO_LED},
+    { 11,  12,  13,  14 },
+    { 15,  NO_LED, NO_LED, 16 }
+}, {
+    {45,  0  },  //  0  NUM
+    {90,  0  },  //  1  /
+    {134, 0  },  //  2  *
+    {179, 0  },  //  3  -
+    {45,  16 },  //  4  7
+    {90,  16 },  //  5  8
+    {134, 16 },  //  6  9
+    {179, 24 },  //  7  +
+    {45,  32 },  //  8  4
+    {90,  32 },  //  9  5
+    {134, 32 },  //  10 6
+    {45,  48 },  //  11 1
+    {90,  48 },  //  12 2
+    {134, 48 },  //  13 3
+    {179, 56 },  //  14 ENTER
+    {67,  64 },  //  15 0
+    {134, 64 },  //  16 .
+    {0,   0  },  //  17 LED18
+    {0,   11 },  //  18 LED19
+    {0,   21 },  //  19 LED20
+    {0,   32 },  //  20 LED21
+    {0,   43 },  //  21 LED22
+    {0,   53 },  //  22 LED23
+    {0,   64 },  //  23 LED24
+    {224, 0  },  //  24 LED27
+    {224, 11 },  //  25 LED28
+    {224, 21 },  //  26 LED29
+    {224, 32 },  //  27 LED30
+    {224, 43 },  //  28 LED31
+    {224, 53 },  //  29 LED32
+    {224, 64 }   //  30 LED33
+}, {
+    4, 4, 4, 4,
+    4, 4, 4, 4,
+    4, 4, 4,
+    4, 4, 4, 4,
+    4,       4,
+    2, 2, 2, 2, 2, 2, 2,
+	2, 2, 2, 2, 2, 2, 2
+} };
+
+#    ifdef DRIVER_1_PW_EN
+
+void matrix_init_kb(void) {
+    wait_ms(2000);
+    setPinOutput(DRIVER_1_PW_EN);
+    writePinHigh(DRIVER_1_PW_EN);
+    matrix_init_user();
+}
```

Comment:
```suggestion
void keyboard_post_init_kb(void) {
    wait_ms(2000);
    setPinOutput(DRIVER_1_PW_EN);
    writePinHigh(DRIVER_1_PW_EN);
    keyboard_post_init_user();
}
```

Also, is the wait needed, especially with the post init?

Reply:
Not really sure, I can test and investigate in the morning to see if it is needed.

Reply:
shouldn't this code be in `keyboard_pre_init_user` as per the documentation? The doc says led initialization should be there.
 https://github.com/qmk/qmk_firmware/blob/master/docs/custom_quantum_functions.md#example-keyboard_pre_init_user-implementation

Reply:
yeah, because wait_ms is hard stop (eg, nothing else runs).  So if it can be removed, that would be idea.

Reply:
@drashna removing the `wait_ms(200);` breaks RGB, moving it to `keyboard_pre_init_user()` breaks RGB, moving it to `keyboard_post_init_kb()` also breaks RGB

Seems like the only way RGB still works is by keeping it in the matrix init code... executing it earlier than matrix breaks the RGB

Reply:
@drashna I can move it to `keyboard_pre_init_user()` but it still requires the `wait_ms(200);`

Reply:
A 2 second delay at startup is really unfortunate. At least it's a numpad, and not a regular keyboard.

Apparently `matrix_init()` is about the last replaceable hook that is invoked before `rgb_matrix_init()` (in theory there is also `led_init_ports()`, but that's for a completely different feature, so its usage may be considered as a dirty hack; and the only user-visible change would be that the 2 second delay would happen after the bootmagic check).

I wonder what is the underlying problem here — no sane chip would require such delays (the AW20216S datasheet mentions that a 2 ms delay is required after a power-on reset, but that's 1000 times lower than what is done here).

Reply:
@sigprof I never tried to use a faster wait time, it could be that @GloriousThrall read 2 seconds which would be 2000 ms instead of 2 ms

Reply:
If it was the case, it would have worked without the delay, because the USB init code has a 50 ms delay in it — but as you observe that RGB does not work without that delay, there is something else there that requires a long delay at startup.

Reply:
I got this lowered to 1000ms, when I try to go lower than this it breaks RGB

Reply:
Without any reliable information about why that delay is needed you may need to keep it at 2000 ms (maybe 1000 ms is enough for your particular board, but not in general).

Maybe the problem is caused by the wireless part (which is not touched by this code) — e.g., there might be some circuit which switches off the power to the LED controller.

BTW, one thing you can do here is to put the code inside `#ifdef RGB_MATRIX_ENABLE`, so that the delay won't be applied if someone decides to turn off RGB support completely.

Reply:
@sigprof so I should change `#ifdef DRIVER_1_PW_EN` to `#ifdef RGB_MATRIX_ENABLE`

Reply:
It looks like the entire RGB matrix along with this trouble code is already wrapped in `RGB_MATRIX_ENABLE`

Reply:
Ah, then that's OK as is — did not look at the full file.